### PR TITLE
render() and check() refactor to keep their argument to the minumum needed

### PR DESCRIPTION
**Pull request by fgalan**
This PR is mainly a simplification of the render() and check() method signatures for classes in libraries ngsi/, ngsi10/, ngsi9/, orionTypes/ and apiTypesV2/, improving modularity and decoupling. In particular:

* ConnectionInfo module has been decoupled. Only the needed parameters are passed to the methods (most of then in parametes, only a couple of methods use out parameters), not the whole ConnectionInfo. Thus, the signature of the method is now a clear specification of what the method exactly needs to work.
* RequestType arguments removed in ngsi9/ and ngsi10/ classes. The request type itself is provided by the class itself, no need to overload check() and render() with it.

_How to read this PR:_

* Most important: "no functional test were harmed in the filming of this PR" :) This ensures the refactor keeps 100% functional regression.
* Looks only the .h files. The .cpp are not very important in this PR.

### CODE REVIEW

**Comment by crbrox** [comment]
LGTM

File: src/lib/convenience/ContextAttributeResponseVector.cpp [hunk 1]
```diff
@@ -69,18 +74,19 @@ std::string ContextAttributeResponseVector::render(ConnectionInfo* ciP, RequestT
...
 (
-  ConnectionInfo*  ciP,
-  RequestType      request,
-  std::string      indent,
-  std::string      predetectedError,
-  int              counter
+  const std::string&  apiVersion,
+  bool                asJsonObject,
+  RequestType         request,
+  std::string         indent,
```
Comment (kzangeli) [inline src/lib/convenience/ContextAttributeResponseVector.cpp]: Now that we change, let's try to change this as well:

    const std::string&   (x2)

Comment (kzangeli) [inline src/lib/convenience/ContextAttributeResponseVector.cpp]: indent to be left alone (will soon disappear).
predetectedError to be fixed in a following PR.

NTC

File: src/lib/convenience/UpdateContextElementRequest.cpp [hunk 2]
```diff
@@ -83,8 +84,8 @@ std::string UpdateContextElementRequest::check
   if (predetectedError != "")
   {
     response.errorCode.fill(SccBadRequest, predetectedError);
-  }
-  else if ((res = contextAttributeVector.check(ciP, UpdateContextElement, indent, predetectedError, counter)) != "OK")
+  }
```
Comment (kzangeli) [inline src/lib/convenience/UpdateContextElementRequest.cpp]: Extra spaces - style guide checker will report this, once active

File: src/lib/jsonParse/jsonAppendContextElementRequest.cpp [hunk 3]
```diff
@@ -329,7 +329,9 @@ void jsonAcerRelease(ParseData* reqData)
 */
 std::string jsonAcerCheck(ParseData* reqData, ConnectionInfo* ciP)
 {
-  return reqData->acer.res.check(ciP, AppendContextElement, "", reqData->errorString, 0);
+  return reqData->acer.res.check(ciP->apiVersion,
+                                 ciP->uriParam[URI_PARAM_ATTRIBUTE_FORMAT] == "object" && ciP->outMimeType == JSON,
+                                 AppendContextElement, "", reqData->errorString, 0);
 }
```
Comment (kzangeli) [inline src/lib/jsonParse/jsonAppendContextElementRequest.cpp]: Two things here, if the function call is divided into lines, lets give each parameter a line each. Easier to read that way.
Secondly, the second parameter is a bit complex. A help-variable with a descriptive name would make it much easier to understand. E.g.:

    bool jsonObject = (ciP->uriParam[URI_PARAM_ATTRIBUTE_FORMAT] == "object") && (ciP->outMimeType == JSON);

File: src/lib/jsonParseV2/jsonRequestTreat.cpp [hunk 4]
```diff
@@ -74,10 +74,10 @@ std::string jsonRequestTreat
       return answer;
     }
 
-    if ((answer = parseDataP->ent.res.check(ciP, EntitiesRequest)) != "OK")
+    if ((answer = parseDataP->ent.res.check(ciP->apiVersion, EntitiesRequest)) != "OK")
     {
       OrionError oe(SccBadRequest, answer);
-      return oe.setStatusCodeAndSmartRender(ciP);
+      return oe.setStatusCodeAndSmartRender(ciP->apiVersion, &(ciP->httpStatusCode));
```
Comment (kzangeli) [inline src/lib/jsonParseV2/jsonRequestTreat.cpp]: () in &(ciP->httpStatusCode) are not necessary.

However, this is a question for the style-guide.

Personally I find it ugly (I am not used) but at the same time I understand it makes the thing easier to understand (for people that aren't aware of the operator precedence of C).

Let's add it to 'open questions' for the style guide.

NTC

File: src/lib/ngsi10/UpdateContextRequest.cpp [hunk 5]
```diff
@@ -68,16 +66,16 @@ UpdateContextRequest::UpdateContextRequest(const std::string& _contextProvider,
...
-std::string UpdateContextRequest::render(ConnectionInfo* ciP, RequestType requestType, const std::string& indent)
+std::string UpdateContextRequest::render(const std::string& apiVersion, bool asJsonObject, const std::string& indent)
 {
   std::string  out = "";
   std::string  tag = "updateContextRequest";
 
   // JSON commas:
   // Both fields are MANDATORY, so, comma after "contextElementVector"
-  //
+  //
```
Comment (kzangeli) [inline src/lib/ngsi10/UpdateContextRequest.cpp]: trailing spaces

File: src/lib/ngsi9/DiscoverContextAvailabilityRequest.cpp [hunk 6]
```diff
@@ -76,7 +76,7 @@ std::string DiscoverContextAvailabilityRequest::check(ConnectionInfo* ciP, Reque
   {
     response.errorCode.fill(SccContextElementNotFound);
   }
-  else if (((res = entityIdVector.check(ciP, DiscoverContextAvailability, indent, predetectedError, restrictions))                      != "OK") ||
+  else if (((res = entityIdVector.check(DiscoverContextAvailability, indent))                                                      != "OK") ||
```
Comment (kzangeli) [inline src/lib/ngsi9/DiscoverContextAvailabilityRequest.cpp]: spaces

File: src/lib/ngsi9/RegisterContextRequest.cpp [hunk 7]
```diff
@@ -82,8 +82,8 @@ std::string RegisterContextRequest::check(ConnectionInfo* ciP, RequestType reque
   {
     alarmMgr.badInput(clientIp, "empty contextRegistration list");
     response.errorCode.fill(SccBadRequest, "Empty Context Registration List");
-  }
-  else if (((res = contextRegistrationVector.check(ciP, RegisterContext, indent, predetectedError, counter)) != "OK") ||
+  }
```
Comment (kzangeli) [inline src/lib/ngsi9/RegisterContextRequest.cpp]: space

File: src/lib/ngsi9/UpdateContextAvailabilitySubscriptionRequest.cpp [hunk 8]
```diff
@@ -49,11 +49,11 @@ UpdateContextAvailabilitySubscriptionRequest::UpdateContextAvailabilitySubscript
 *
 * UpdateContextAvailabilitySubscriptionRequest::render - 
 */
-std::string UpdateContextAvailabilitySubscriptionRequest::render(RequestType requestType, const std::string& indent)
-{
+std::string UpdateContextAvailabilitySubscriptionRequest::render(const std::string& indent)
+{
```
Comment (kzangeli) [inline src/lib/ngsi9/UpdateContextAvailabilitySubscriptionRequest.cpp]: Last comment about trailing spaces

File: src/lib/serviceRoutines/getEntityTypes.cpp [hunk 9]
```diff
@@ -78,7 +78,11 @@ std::string getEntityTypes
   TIMED_MONGO(mongoEntityTypes(&response, ciP->tenant, ciP->servicePathV, ciP->uriParam, ciP->apiVersion, totalTypesP, true));
 
   std::string rendered;
-  TIMED_RENDER(rendered = response.render(ciP, ""));
+  TIMED_RENDER(rendered = response.render(ciP->apiVersion,
```
Comment (kzangeli) [inline src/lib/serviceRoutines/getEntityTypes.cpp]: Lots and lots of very long lines inside TIMED_RENDER.
Easily fixed using help vars.

Just a suggestion ...

File: src/lib/jsonParseV2/jsonRequestTreat.cpp (hunk 4 above)
Comment (fgalan) [inline src/lib/jsonParseV2/jsonRequestTreat.cpp]: I agree it is ugly but easier to read... and easiness to read wins uglyness :)

File: src/lib/serviceRoutinesV2/getEntityAttributeValue.cpp [hunk 10]
```diff
@@ -103,13 +104,21 @@ std::string getEntityAttributeValue
...
+      TIMED_RENDER(answer = attribute.render(ciP->apiVersion,
+                                             ciP->httpHeaders.accepted("text/plain"),
+                                             ciP->httpHeaders.accepted("application/json"),
+                                             ciP->httpHeaders.outformatSelect(),
+                                             &(ciP->outMimeType),
+                                             &(ciP->httpStatusCode),
+                                             ciP->uriParamOptions[OPT_KEY_VALUES],
+                                             ciP->uriParam[URI_PARAM_METADATA],
+                                             EntityAttributeValueRequest, false));
     }
```
Comment (kzangeli) [inline src/lib/serviceRoutinesV2/getEntityAttributeValue.cpp]: Look at this, a very very long TIMED_RENDER line, and yet, **so** easy to read.
Each param on a separate line and no param excessively long.
Just the way it should be!

Ah, except the last `false` ... :-(

File: src/lib/ngsi9/DiscoverContextAvailabilityRequest.cpp (hunk 6 above)
Comment (fgalan) [inline src/lib/ngsi9/DiscoverContextAvailabilityRequest.cpp]: As discussed at skype, let's consider style-check aspect in the last PR in the "refactor cycle". NTC

File: src/lib/ngsi10/UpdateContextRequest.cpp (hunk 5 above)
Comment (fgalan) [inline src/lib/ngsi10/UpdateContextRequest.cpp]: As discussed at skype, let's consider style-check aspect in the last PR in the "refactor cycle". NTC

File: src/lib/ngsi9/RegisterContextRequest.cpp (hunk 7 above)
Comment (fgalan) [inline src/lib/ngsi9/RegisterContextRequest.cpp]: As discussed at skype, let's consider style-check aspect in the last PR in the "refactor cycle". NTC

File: src/lib/ngsi9/UpdateContextAvailabilitySubscriptionRequest.cpp (hunk 8 above)
Comment (fgalan) [inline src/lib/ngsi9/UpdateContextAvailabilitySubscriptionRequest.cpp]: As discussed at skype, let's consider style-check aspect in the last PR in the "refactor cycle". NTC

File: src/lib/serviceRoutinesV2/getEntityAttributeValue.cpp (hunk 10 above)
Comment (fgalan) [inline src/lib/serviceRoutinesV2/getEntityAttributeValue.cpp]: Fixed in 93eb7c42666230bf5d0996975a85ef5ea5629fc2

File: src/lib/serviceRoutines/getEntityTypes.cpp (hunk 9 above)
Comment (fgalan) [inline src/lib/serviceRoutines/getEntityTypes.cpp]: Fixed in 93eb7c42666230bf5d0996975a85ef5ea5629fc2

File: src/lib/jsonParse/jsonQueryContextRequest.cpp [hunk 11]
```diff
@@ -767,11 +768,8 @@ void jsonQcrRelease(ParseData* reqDataP)
 */
 std::string jsonQcrCheck(ParseData* reqDataP, ConnectionInfo* ciP)
 {
-  return reqDataP->qcr.res.check(ciP,
-                                 DiscoverContextAvailability,
-                                 "",
-                                 reqDataP->errorString,
-                                 reqDataP->qcr.res.restrictions);
+  return reqDataP->qcr.res.check(ciP->apiVersion,
+                                 ciP->uriParam[URI_PARAM_ATTRIBUTE_FORMAT] == "object" && ciP->outMimeType == JSON, "", reqDataP->errorString);
```
Comment (fgalan) [inline src/lib/jsonParse/jsonQueryContextRequest.cpp]: Fixed in 93eb7c42666230bf5d0996975a85ef5ea5629fc2

File: src/lib/jsonParseV2/parseBatchQuery.cpp [hunk 12]
```diff
@@ -104,7 +104,7 @@ std::string parseBatchQuery(ConnectionInfo* ciP, BatchQuery* bqrP)
       std::string r = parseAttributeList(ciP, iter, &bqrP->attributeV);
 
       if (r != "OK")
-      {        
+      {
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseBatchQuery.cpp]: Not actually related with this PR... but I spotted (this an some similar ones) during last commit, so I fixed.

NTC

**Comment by fgalan** [comment]
I have reviewed all the ~130 ocurrences of TIMED_RENDER() and all them use `asJsonObject` bool variable to avoid a long logical sentence and use the one-argument-per-line style in the case of multinline (only a couple of cases at the end).

I think the PR is reaady now for re-review and merge.

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**